sdk/ml/azure-ai-ml/tests/dsl/unittests/test_dsl_pipeline.py

```diff
@@ -2086,3 +2086,51 @@ def pipeline_func(job_in_number, job_in_path, component_func):
             component_func=component_func2)
         assert len(pipeline.jobs) == 2
         assert component_func2.name in pipeline.jobs
+
+    def test_pipeline_with_variable_inputs(self):
+        path = "./tests/test_configs/components/helloworld_component.yml"
+        component_func1 = load_component(path)
+        data = Data(name="test", version="1", type=AssetTypes.MLTABLE)
+
+        @dsl.pipeline
+        def pipeline_with_variable_args(*args, **kwargs):
+            node_args = component_func1(component_in_number=args[0], component_in_path=args[1])
+            node_kwargs = component_func1(component_in_number=kwargs["component_in_number1"],
+                                          component_in_path=kwargs["component_in_path1"])
+
+        @dsl.pipeline
+        def root_pipeline(component_in_number: int, component_in_path: Input, *args, **kwargs):
+            node = component_func1(component_in_number=component_in_number, component_in_path=component_in_path)
+            node_args = component_func1(component_in_number=args[0], component_in_path=args[1])
+            node_kwargs = component_func1(component_in_number=kwargs["component_in_number1"],
+                                          component_in_path=kwargs["component_in_path1"])
+            node_with_arg_kwarg = pipeline_with_variable_args(*args, **kwargs)
+
+        pipeline = root_pipeline(10, data, 11, data, component_in_number1=11, component_in_path1=data)
+
+        assert len(pipeline.inputs) == 6
+        arg_names = ['component_in_number', 'component_in_path', 'component_in_number1',
+                     'component_in_path1', 'args_0', 'args_1']
+        assert all([input_name in pipeline.inputs for input_name in arg_names])
+
+        node_args = pipeline.jobs['node_args']
+        assert len(node_args.inputs) == 2
+        assert all([input._data._name in ['args_0', 'args_1'] for input in node_args.inputs.values()])
+
+        node_kwargs = pipeline.jobs['node_kwargs']
+        assert len(node_kwargs.inputs) == 2
+        assert all([input._data._name in ['component_in_number1', 'component_in_path1'] for input in
+                    node_kwargs.inputs.values()])
+
+        node_with_arg_kwarg = pipeline.jobs['node_with_arg_kwarg']
+        assert len(node_with_arg_kwarg.inputs) == 4
+        arg_names = ['component_in_number1', 'component_in_path1', 'args_0', 'args_1']
+        assert all([input._data._name in arg_names for input in node_with_arg_kwarg.inputs.values()])
+
+        with mock.patch(
+            "azure.ai.ml.entities._job.pipeline.pipeline_job.is_private_preview_enabled",
+            return_value=True
+        ):
+            with pytest.raises(UnsupportedParameterKindError,
```

Comment:
what's the behavior if we do mldesigenr compile

Reply:
If the user indexes by index or key in the pipeline code, it will raise error when execute pipeline code.
```
Traceback (most recent call last):
  File "d:\project\azure-sdk-for-python\sdk\ml\azure-ai-ml\azure\ai\ml\dsl\_pipeline_component_builder.py", line 215, in build
    outputs = self.func(*pipeline_args, **pipeline_kwargs)
  File "D:/Project/azure-sdk-for-python/sdk/ml/azure-ai-ml/tests/test.py", line 50, in root_pipeline
    node_args = component_func1(component_in_number=args[0], component_in_path=args[1])
IndexError: tuple index out of range
```
```
Traceback (most recent call last):
  File "d:\project\azure-sdk-for-python\sdk\ml\azure-ai-ml\azure\ai\ml\dsl\_pipeline_component_builder.py", line 215, in build
    outputs = self.func(*pipeline_args, **pipeline_kwargs)
  File "D:/Project/azure-sdk-for-python/sdk/ml/azure-ai-ml/tests/test.py", line 51, in root_pipeline
    node_kwargs = component_func1(component_in_number=kwargs["component_in_number1"],
KeyError: 'component_in_number1'
```

Reply:
could we raise error before execute user's code?